### ISSUE TITLE
Stackage

### ISSUE THREAD

**Issue by fakedrake** [svenkeidel/sturdy#27]
Could we get a package out of this? I like your code on arrows

**Comment by svenkeidel** [comment]
Thanks for checking out my library. Yes, this sounds like a good idea. @seba--, what do you think?

I'm trying to decide if I should split this library into smaller packages or if I should upload it as a whole. @fakedrake, which part of the library would you use? Would you use the library for implementing static analyses or would you only use the arrow code for implementing other applications?

**Comment by fakedrake** [comment]
The overall issue I am facing is that there is no arrow transformer library on stackage so I have to re-implement everything. Which sucks. So a package with everything under `Control.Arrow` would be perfect.

**Comment by svenkeidel** [comment]
Makes sense. Well, there is the arrow transformer library [atl](https://hackage.haskell.org/package/atl). Have you tried atl and is there a problem with it?

Furthermore, currently half of whats in `sturdy:Control.Arrow` is specific to static analyses and this is probably of no use for the day to day developer. What would be of use are the general purpose transformers `ReaderT`, `WriterT`, `StateT`, `ConstT`, `StaticT`, `ContT`, `KleisliT`, and `CoKleisliT`. These transformers could be extract into their own package, however, I need to be convinced that this is worth the effort. Maybe it is worth to contribute some of the arrow code to the atl project.

**Comment by svenkeidel** [comment]
[Here](https://github.com/svenkeidel/sturdy/files/4717488/library.txt) is a list of all the files that are general purpose and reusable. I removed the `ExceptT`-like transformers because they include some details specific to static analysis.

**Comment by fakedrake** [comment]
The list looks legit. TBH I only shamelessly copied `KleisliT`, `StateT`, `ReaderT` and `ContT` (so far). Better than atl is [arrows](https://hackage.haskell.org/package/arrows) but it's not on stackage, it seems discontinued, it's missing important transformers (ehm `KleisliT`), and most importantly I like the implementations in sturdy much better.

**Comment by svenkeidel** [comment]
Ok, maybe its time for a new arrow library. I see what I can do, but this may take some time because I have to work on this in my spare time.

**Comment by seba--** [comment]
I think this is a great idea.

**Comment by svenkeidel** [comment]
@fakedrake, you can follow the current progress at https://github.com/svenkeidel/sturdy/tree/extract-arrow-transformer-library.

**Comment by fakedrake** [comment]
Wow you did quite a bit of work, thanks! Anyway documentation looks legit. Thanks for moving some benchmarks in there too. 

What's up with changing around `ArrowTrans` and `ArrowLift`?

**Comment by fakedrake** [comment]
This is low priority by I like [Automaton](https://hackage.haskell.org/package/arrows-0.4.4.2/docs/Control-Arrow-Transformer-Automaton.html), if you ever get the chance to add it. [This](https://hackage.haskell.org/package/machines-0.7/docs/Data-Machine-Mealy.html) implementation has better naming and implementation (IMHO) but it's not a transformer.

**Comment by svenkeidel** [comment]
> What's up with changing around ArrowTrans and ArrowLift ?

I swapped the names to make the name of `ArrowTrans` more consistent with mtl's `MonadTrans`. The `ArrowTrans.lift'` function is the arrow equivalent of `MonadTrans.lift`.

> This is low priority by I like Automaton

Sure I see what I can do. I have never used this transformer. What is it useful for?

**Comment by fakedrake** [comment]
One example might be [circuits](https://ieeexplore.ieee.org/abstract/document/6037453). I have seen in for stream processing but I can't find the citation right now.

**Comment by fakedrake** [comment]
I saw you added it, thanks a lot! Final request and I will leave you alone: could you move `ExceptT` in there as well?

**Comment by svenkeidel** [comment]
I see. So you want to eliminate the `ArrowMonad` class and instead have a `Category Kleisli` instance vor every Monad. I like this idea.

About adding FailureT: the only problem I have is that the `ArrowFail` class is too general due to it's use for static analysis. Either we use this more general version or introduce a simpler version without the `Join` constraint.

**Comment by fakedrake** [comment]
Sorry, I am realizing that what I wrote was not clear at all. What I mean is this: 

```
newtype ExceptT e c a b = ExceptT (c a (Either e b))
instance ArrowChoice c => Category (ExceptT e c) where
  ExceptT f . ExceptT g = ExceptT $ g >>> arr Left ||| f
  id = ExceptT $ arr return . id

instance ArrowChoice c => Arrow (ExceptT f c) where
  arr f = ExceptT $ arr $ return . f
  first (ExceptT f) = ExceptT $ first f >>> arr (\(fc,u) -> (,u) <$> fc)
```

The problem is that as far as I can tell, the only `ArrowMonad` is `KleisliT` and the arrows implemented as such, but almost all the arrows in the package are `ArrowChoice`.

I know performance is important for sturdy so disregard this if this change affects performance badly.
